components/ratgdo/ratgdo.cpp

```diff
@@ -485,28 +595,127 @@ namespace ratgdo {
+        }
+        auto updates = this->openingDuration * 1000 * delta/update_period;
+        auto d = delta/updates;
+        auto count = int(updates);
+        ESP_LOGD(TAG, "[Opening] Position sync %d times: ", count);
+        // try to keep position in sync while door is moving
+        set_retry("position_sync_while_moving", update_period, count, [=](uint8_t r) {
+            ESP_LOGD(TAG, "[Opening] Position sync: %d: ", r);
+            this->doorPosition += d;
+            return RetryResult::RETRY;
+        });
+    }
+
+    void RATGDOComponent::positionSyncWhileClosing(float delta, float update_period) {
+        if (this->closingDuration==0) {
+            ESP_LOGW(TAG, "I don't know closing duration, ignoring position sync");
+            return;
+        }
+        auto updates = this->closingDuration * 1000 * delta/update_period;
+        auto d = delta/updates;
+        auto count = int(updates);
+        ESP_LOGD(TAG, "[Closing] Position sync %d times: ", count);
+        // try to keep position in sync while door is moving
+        set_retry("position_sync_while_moving", update_period, count, [=](uint8_t r) {
+            ESP_LOGD(TAG, "[Closing] Position sync: %d: ", r);
+            this->doorPosition -= d;
+            return RetryResult::RETRY;
+        });
+    }
+
+    void RATGDOComponent::setDoorPosition(float position)
+    {
+        if (this->doorState == DoorState::DOOR_STATE_OPENING || this->doorState == DoorState::DOOR_STATE_CLOSING) {
+            ESP_LOGW(TAG, "The door is moving, ignoring.");
+            return;
+        }
+
+        auto delta = position - this->doorPosition;
+        if (delta > 0) { // open
+            if (this->openingDuration==0) {
+                ESP_LOGW(TAG, "I don't know opening duration, ignoring move to position");
+                return;
+            }
+            auto opening_time = this->openingDuration * 1000 * delta;
+            doorCommand(data::DOOR_OPEN);
+            this->movingToPosition = true;
+            set_timeout("move_to_position", opening_time, [=] {
+                doorCommand(data::DOOR_STOP);
+                this->movingToPosition = false;
+                this->doorPosition = position;
+            });
+            this->positionSyncWhileOpening(delta);
+        } else if (delta < 0) { // close
+            if (this->closingDuration==0) {
+                ESP_LOGW(TAG, "I don't know closing duration, ignoring move to position");
+                return;
+            }
+            delta = -delta;
+            auto closing_time = this->closingDuration * 1000 * delta;
+            doorCommand(data::DOOR_CLOSE);
+            this->movingToPosition = true;
+            set_timeout("move_to_position", closing_time, [=] {
+                doorCommand(data::DOOR_STOP);
+                this->movingToPosition = false;
+                this->doorPosition = position;
+            });
+            this->positionSyncWhileClosing(delta);
+        }
+    }
```

Comment:
There is a bit of duplication here. It would be nice if this was a bit more DRY

Reply:
Indeed, but removing the duplication makes the code quite a bit less readable I think.

Reply:
https://github.com/ESPHome-RATGDO/esphome-ratgdo/pull/21

I think `set_timeout` was being called with a negative value when closing

Reply:
nah I read it wrong when I refactored